{
 "cells": [
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "1. Split MIBiG compounds into 5 groups for cross-validation based on BigScape families.",
   "id": "35f84ba0fa33b30e"
  },
  {
   "metadata": {
    "collapsed": true,
    "ExecuteTime": {
     "end_time": "2025-06-11T13:09:21.571311Z",
     "start_time": "2025-06-11T13:09:21.346518Z"
    }
   },
   "cell_type": "code",
   "source": [
    "from math import ceil\n",
    "from pathlib import Path\n",
    "import pandas as pd\n",
    "\n",
    "nerpa_dir = Path('/home/ilianolhin/git/nerpa2')\n",
    "\n",
    "mibig_bgcs_table = pd.read_csv(nerpa_dir / 'scripts/build_mibig_info_table/mibig_bgcs_info.tsv', sep='\\t')\n",
    "mibig_compounds_approved = mibig_bgcs_table[mibig_bgcs_table[\"in_approved_matches\"] == True].copy()\n",
    "\n",
    "if mibig_compounds_approved[\"bigscape_families\"].isna().sum() > 0:\n",
    "    print('Warning! Missing bigscape_families values present.')\n",
    "\n",
    "# Aggregate BGCs by BigScape family using groupby\n",
    "family_to_bgcs = mibig_compounds_approved.groupby(\"bigscape_families\")[\"bgc_id\"].apply(list).to_dict()\n",
    "\n",
    "n_groups = 5  # Number of groups to split the families into\n",
    "total_bgcs = sum(len(bgcs) for bgcs in family_to_bgcs.values())\n",
    "target_group_size = ceil(total_bgcs / n_groups)\n",
    "\n",
    "family_iter = iter(family_to_bgcs.items())\n",
    "family_to_sample_group = {}\n",
    "\n",
    "for group_idx in range(n_groups):\n",
    "    group_size = 0\n",
    "    while group_size < target_group_size and (family := next(family_iter, None)) is not None:\n",
    "        family_name, bgcs = family\n",
    "        family_to_sample_group[family_name] = group_idx\n",
    "        group_size += len(bgcs)\n",
    "\n",
    "# q: assert that all families are assigned to a group\n",
    "assert len(family_to_sample_group) == len(family_to_bgcs), \"Not all families are assigned to a group\"\n",
    "\n",
    "# Assign sample group to each row\n",
    "mibig_compounds_approved[\"sample_group\"] = (mibig_compounds_approved[\"bigscape_families\"]\n",
    "                                            .map(family_to_sample_group)\n",
    "                                            .astype('Int64'))\n",
    "\n",
    "# display mibig_compounds_approved\n",
    "print(mibig_compounds_approved.head())"
   ],
   "id": "initial_id",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning! Missing bigscape_families values present.\n",
      "        bgc_id  num_a_domains origins bigscape_families   compound_id  \\\n",
      "15  BGC0000296              6  mibig3          NRPS_277  BGC0000296.0   \n",
      "27  BGC0000305             11  mibig3  NRPS_285_mix_285  BGC0000305.0   \n",
      "28  BGC0000306              6  mibig3  NRPS_286_mix_286  BGC0000306.0   \n",
      "29  BGC0000307              9  mibig3  NRPS_287_mix_287  BGC0000307.0   \n",
      "34  BGC0000310             12  mibig3  NRPS_290_mix_290  BGC0000310.0   \n",
      "\n",
      "    num_recognized_nodes  iso_class_idx  in_approved_matches  sample_group  \n",
      "15                    11           3006                 True             3  \n",
      "27                    11           1782                 True             3  \n",
      "28                     6            368                 True             3  \n",
      "29                     9           3138                 True             3  \n",
      "34                    12            612                 True             3  \n"
     ]
    }
   ],
   "execution_count": 1
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "2. Form subsets of approved matches with MIBiG compounds for cross-validation.",
   "id": "d52c0d32b99d44b7"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-11T13:27:28.987721Z",
     "start_time": "2025-06-11T13:27:28.443567Z"
    }
   },
   "cell_type": "code",
   "source": [
    "from itertools import chain\n",
    "from src.matching.match_type import Match\n",
    "approved_matches_txt = Path(nerpa_dir / 'test_data/approved_matches/approved_matches.txt')\n",
    "approved_matches_strs = approved_matches_txt.read_text().split('\\n\\n')\n",
    "approved_matches = (Match.from_str(match_str)\n",
    "                    for match_str in approved_matches_strs\n",
    "                    if match_str.strip())\n",
    "approved_matches_per_compound = {match.nrp_variant_id.nrp_id: match\n",
    "                                 for match in approved_matches}\n",
    "\n",
    "benchmarking_dir = Path(nerpa_dir / 'benchmarking')\n",
    "(benchmarking_dir / 'approved_matches_subsets').mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "training_subset_paths = [benchmarking_dir / 'approved_matches_subsets' / f'training_subset_{i}.txt'\n",
    "                         for i in range(n_groups)]\n",
    "testing_subset_paths = [benchmarking_dir / 'approved_matches_subsets' / f'testing_subset_{i}.txt'\n",
    "                        for i in range(n_groups)]\n",
    "\n",
    "subsets = mibig_compounds_approved.groupby('sample_group')['compound_id'].apply(list).to_dict()\n",
    "\n",
    "\"\"\"\n",
    "assert set(approved_matches_per_compound.keys()) == set(chain(*subsets.values())), \\\n",
    "    (\"The set of approved matches does not match the set of MIBiG compounds in subsets.\\n\"\n",
    "    f\"subsets \\ approved_matches = {set(chain(*subsets.values())) - set(approved_matches_per_compound.keys())}\\n\"\n",
    "    f\"approved_matches \\ subsets = {set(approved_matches_per_compound.keys()) - set(chain(*subsets.values()))}\")\n",
    "\"\"\"\n",
    "\n",
    "for group_idx, compound_ids in subsets.items():\n",
    "    testing_subset = [approved_matches_per_compound[nrp_id]\n",
    "                      for nrp_id in compound_ids]\n",
    "    training_subset = [match\n",
    "                       for match in approved_matches_per_compound.values()\n",
    "                       if match.nrp_variant_id.nrp_id not in compound_ids]\n",
    "\n",
    "    # Save the subset to a file\n",
    "    with open(testing_subset_paths[group_idx], 'w') as f:\n",
    "        f.write('\\n'.join(match.nrp_variant_id.nrp_id\n",
    "                          for match in testing_subset))\n",
    "\n",
    "    with open(training_subset_paths[group_idx], 'w') as f:\n",
    "        f.write('\\n\\n'.join(map(str, training_subset)))"
   ],
   "id": "891581b93b6d2240",
   "outputs": [],
   "execution_count": 7
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "3. Create a config file for each subset.",
   "id": "bbf3473d65c6ec79"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-11T13:14:43.681462Z",
     "start_time": "2025-06-11T13:11:41.263502Z"
    }
   },
   "cell_type": "code",
   "source": [
    "import subprocess\n",
    "(benchmarking_dir / 'training_results').mkdir(parents=True, exist_ok=True)\n",
    "training_results_paths = [benchmarking_dir / 'training_results' / f'subset_{i}'\n",
    "                          for i in range(n_groups)]\n",
    "for subset_idx in range(n_groups):\n",
    "    command = (f'python {nerpa_dir / \"train_nerpa.py\"} '\n",
    "               f'--approved-matches {training_subset_paths[subset_idx]} '\n",
    "               f'--output-dir {training_results_paths[subset_idx]}')\n",
    "    print(f'Running command: {command}')\n",
    "    subprocess.run(command, shell=True, check=True)"
   ],
   "id": "f22070ca04b7afae",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running command: python /home/ilianolhin/git/nerpa2/train_nerpa.py --approved-matches /home/ilianolhin/git/nerpa2/benchmarking/approved_matches_subsets/training_subset_0.txt --output-dir /home/ilianolhin/git/nerpa2/benchmarking/training_results/subset_0\n",
      "Loading approved matches\n",
      "Loading BGC variants\n",
      "Inferring edge parameters from counts...\n",
      "WARNING(ok): not enough data for INSERT_AT_START, so I use the same as for INSERT\n",
      "WARNING(ok): not enough data for influence of PKS_DOWNSTREAM on START_INSERTING, so I ignore it\n",
      "WARNING(ok): not enough data for influence of PKS_DOWNSTREAM on START_INSERTING_AT_END, so I ignore it\n",
      "WARNING: Weights for edge type DetailedHMMEdgeType.SKIP_GENE and genomic context () not computed - setting to min allowed value: -5\n",
      "Building step function...\n",
      "Calculating modifications frequencies...\n",
      "Calculating PKS probability...\n",
      "Unknown residue predictions correctness: [(0.008, False), (0.008, False), (0.008, False), (0.008, False), (0.008, False), (0.008, False), (0.008, False), (0.008, False), (0.008, True), (0.008, False), (0.008, True), (0.008, True), (0.008, False), (0.008, False), (0.008, False), (0.008, False), (0.008, False), (0.008, False), (0.008, False), (0.008, False), (0.008, False), (0.008, True)]\n",
      "Posterior mean      = 0.2019\n",
      "95 % credible band  = (0.0670, 0.3828)\n",
      "Running command: python /home/ilianolhin/git/nerpa2/train_nerpa.py --approved-matches /home/ilianolhin/git/nerpa2/benchmarking/approved_matches_subsets/training_subset_1.txt --output-dir /home/ilianolhin/git/nerpa2/benchmarking/training_results/subset_1\n",
      "Loading approved matches\n",
      "Loading BGC variants\n",
      "Inferring edge parameters from counts...\n",
      "WARNING(ok): not enough data for INSERT_AT_START, so I use the same as for INSERT\n",
      "WARNING(ok): not enough data for influence of PKS_DOWNSTREAM on START_INSERTING, so I ignore it\n",
      "WARNING(ok): not enough data for influence of PKS_DOWNSTREAM on START_INSERTING_AT_END, so I ignore it\n",
      "WARNING: Weights for edge type DetailedHMMEdgeType.SKIP_GENE and genomic context () not computed - setting to min allowed value: -5\n",
      "Building step function...\n",
      "Calculating modifications frequencies...\n",
      "Calculating PKS probability...\n",
      "Unknown residue predictions correctness: [(0.008, False), (0.008, True), (0.008, False), (0.008, False), (0.008, False), (0.008, False), (0.008, False), (0.008, False), (0.008, False), (0.008, False), (0.008, False), (0.008, False), (0.008, True), (0.008, False), (0.008, False), (0.008, False), (0.008, False), (0.008, False), (0.008, False), (0.008, False), (0.008, False), (0.008, False), (0.008, False), (0.008, True)]\n",
      "Posterior mean      = 0.1470\n",
      "95 % credible band  = (0.0375, 0.3068)\n",
      "Running command: python /home/ilianolhin/git/nerpa2/train_nerpa.py --approved-matches /home/ilianolhin/git/nerpa2/benchmarking/approved_matches_subsets/training_subset_2.txt --output-dir /home/ilianolhin/git/nerpa2/benchmarking/training_results/subset_2\n",
      "Loading approved matches\n",
      "Loading BGC variants\n",
      "Inferring edge parameters from counts...\n",
      "WARNING(ok): not enough data for INSERT_AT_START, so I use the same as for INSERT\n",
      "WARNING(ok): not enough data for influence of PKS_DOWNSTREAM on START_INSERTING, so I ignore it\n",
      "WARNING(ok): not enough data for influence of PKS_DOWNSTREAM on START_INSERTING_AT_END, so I ignore it\n",
      "WARNING: Weights for edge type DetailedHMMEdgeType.SKIP_GENE and genomic context () not computed - setting to min allowed value: -5\n",
      "Building step function...\n",
      "Calculating modifications frequencies...\n",
      "Calculating PKS probability...\n",
      "Unknown residue predictions correctness: [(0.008, True), (0.008, False), (0.008, False), (0.008, False), (0.008, False), (0.008, False), (0.008, False), (0.008, False), (0.008, False), (0.008, False), (0.008, False), (0.008, False), (0.008, True), (0.008, False), (0.008, True), (0.008, True), (0.008, False), (0.008, False), (0.008, False), (0.008, False), (0.008, False), (0.008, False), (0.008, False), (0.008, False), (0.008, False), (0.008, True)]\n",
      "Posterior mean      = 0.2079\n",
      "95 % credible band  = (0.0790, 0.3758)\n",
      "Running command: python /home/ilianolhin/git/nerpa2/train_nerpa.py --approved-matches /home/ilianolhin/git/nerpa2/benchmarking/approved_matches_subsets/training_subset_3.txt --output-dir /home/ilianolhin/git/nerpa2/benchmarking/training_results/subset_3\n",
      "Loading approved matches\n",
      "Loading BGC variants\n",
      "Inferring edge parameters from counts...\n",
      "WARNING(ok): not enough data for INSERT_AT_START, so I use the same as for INSERT\n",
      "WARNING(ok): not enough data for influence of PKS_DOWNSTREAM on START_INSERTING, so I ignore it\n",
      "WARNING(ok): not enough data for influence of PKS_DOWNSTREAM on START_INSERTING_AT_END, so I ignore it\n",
      "WARNING: Weights for edge type DetailedHMMEdgeType.SKIP_GENE and genomic context () not computed - setting to min allowed value: -5\n",
      "Building step function...\n",
      "Calculating modifications frequencies...\n",
      "Calculating PKS probability...\n",
      "Unknown residue predictions correctness: [(0.008, False), (0.008, True), (0.008, False), (0.008, False), (0.008, False), (0.008, False), (0.008, False), (0.008, False), (0.008, False), (0.008, False), (0.008, False), (0.008, False), (0.008, False), (0.008, True), (0.008, False), (0.008, True), (0.008, True), (0.008, False), (0.008, False), (0.008, False), (0.008, False), (0.008, False), (0.008, False), (0.008, False), (0.008, True)]\n",
      "Posterior mean      = 0.2159\n",
      "95 % credible band  = (0.0825, 0.3888)\n",
      "Running command: python /home/ilianolhin/git/nerpa2/train_nerpa.py --approved-matches /home/ilianolhin/git/nerpa2/benchmarking/approved_matches_subsets/training_subset_4.txt --output-dir /home/ilianolhin/git/nerpa2/benchmarking/training_results/subset_4\n",
      "Loading approved matches\n",
      "Loading BGC variants\n",
      "Inferring edge parameters from counts...\n",
      "WARNING(ok): not enough data for INSERT_AT_START, so I use the same as for INSERT\n",
      "WARNING(ok): not enough data for influence of PKS_DOWNSTREAM on START_INSERTING, so I ignore it\n",
      "WARNING(ok): not enough data for influence of PKS_DOWNSTREAM on START_INSERTING_AT_END, so I ignore it\n",
      "WARNING: Weights for edge type DetailedHMMEdgeType.START_INSERTING and genomic context (<ModuleGenomicContextFeature.PKS_DOWNSTREAM: 4>,) not computed - setting to min allowed value: -5\n",
      "WARNING: Weights for edge type DetailedHMMEdgeType.SKIP_GENE and genomic context () not computed - setting to min allowed value: -5\n",
      "Building step function...\n",
      "Calculating modifications frequencies...\n",
      "Calculating PKS probability...\n",
      "Unknown residue predictions correctness: [(0.008, False), (0.008, True), (0.008, False), (0.008, False), (0.008, False), (0.008, False), (0.008, False), (0.008, False), (0.008, False), (0.008, True), (0.008, True), (0.008, False), (0.008, False), (0.008, False), (0.008, False), (0.008, False)]\n",
      "Posterior mean      = 0.2160\n",
      "95 % credible band  = (0.0605, 0.4298)\n"
     ]
    }
   ],
   "execution_count": 6
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "4. Prepare inputs for Nerpa",
   "id": "df59f6316b9f8240"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-11T13:52:45.571932Z",
     "start_time": "2025-06-11T13:52:45.517295Z"
    }
   },
   "cell_type": "code",
   "source": [
    "import shutil\n",
    "\n",
    "inputs_dir = benchmarking_dir / 'cross_validation_inputs'\n",
    "inputs_dir.mkdir(parents=True, exist_ok=True)\n",
    "as_results_on_mibig_dir = Path('/home/ilianolhin/programs/antismash7/as_results_mibig4_nrps')\n",
    "compounds_table = pd.read_csv(nerpa_dir / 'data/pnrpdb2rc1_summary.tsv', sep='\\t')\n",
    "\n",
    "for subset_idx in range(n_groups):\n",
    "    subset_input_dir = inputs_dir / f'subset_{subset_idx}'\n",
    "    # 1. copy configs from training results\n",
    "    source = training_results_paths[subset_idx] / 'new_configs'\n",
    "    dest = subset_input_dir / 'configs'\n",
    "    dest.mkdir(parents=True, exist_ok=True)\n",
    "    shutil.copytree(source, dest, dirs_exist_ok=True)\n",
    "\n",
    "    # 2. Load nrp_ids for testing\n",
    "    nrp_ids = [nrp_id.strip()\n",
    "               for nrp_id in testing_subset_paths[subset_idx].read_text().split('\\n')\n",
    "               if nrp_id.strip()]\n",
    "    bgc_ids = {nrp_id.split('.')[0] for nrp_id in nrp_ids}\n",
    "\n",
    "    # 2. Create a file with antismash paths\n",
    "    antismash_paths = [as_results_on_mibig_dir / bgc_id\n",
    "                for bgc_id in bgc_ids]\n",
    "    with open(subset_input_dir / 'antismash_paths.txt', 'w') as f:\n",
    "        f.write('\\n'.join(str(path) for path in antismash_paths))\n",
    "\n",
    "    # 3. Create a file with MIBiG compounds\n",
    "    # q: assert that all nrp_ids are in compounds_table\n",
    "    assert set(nrp_ids).issubset(set(compounds_table['ID'])), \\\n",
    "        \"Not all nrp_ids are in compounds_table\"\n",
    "\n",
    "    mibig_compounds = compounds_table[compounds_table['ID'].isin(nrp_ids)]\n",
    "    mibig_compounds.to_csv(subset_input_dir / 'compounds.tsv',\n",
    "                           sep='\\t', index=False)\n",
    "\n"
   ],
   "id": "a393a319c68b1ca",
   "outputs": [],
   "execution_count": 11
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-11T15:42:54.252566Z",
     "start_time": "2025-06-11T15:42:53.184929Z"
    }
   },
   "cell_type": "code",
   "source": [
    "def is_from_mibig_norine(nrp_id: str) -> bool:\n",
    "    return nrp_id.startswith('BGC') or nrp_id.startswith('NOR')\n",
    "\n",
    "def at_least_3_recognized_monomers(nrp_id: str) -> bool:\n",
    "    \"\"\"Check if the nrp_id has at least 3 recognized monomers in the compound_stats_table.\"\"\"\n",
    "    values = compound_stats_table[compound_stats_table['compound_id'] == nrp_id]['num_recognized_nodes'].values\n",
    "    return values.size > 0 and values[0] >= 3\n",
    "\n",
    "compound_stats_table = pd.read_csv(nerpa_dir / 'scripts/build_mibig_info_table/rban_graphs_filtered.tsv', sep='\\t')\n",
    "mibig_norine_compounds = (compounds_table[compounds_table['ID']\n",
    "                        .apply(lambda nrp_id: is_from_mibig_norine(nrp_id) and at_least_3_recognized_monomers(nrp_id))])\n",
    "\n",
    "mibig_norine_compounds.to_csv(nerpa_dir / 'data' / 'mibig_norine_compounds.tsv',\n",
    "                               sep='\\t', index=False)"
   ],
   "id": "e93e3a890e2dd5e0",
   "outputs": [],
   "execution_count": 21
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "",
   "id": "f42632320e51a339"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "5. Run Nerpa on each subset.",
   "id": "a34945bf1192f95"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-11T15:58:09.445818Z",
     "start_time": "2025-06-11T15:55:44.043983Z"
    }
   },
   "cell_type": "code",
   "source": [
    "results_dir = benchmarking_dir / 'cross_validation_results'\n",
    "results_dir.mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "for subset_idx in range(n_groups):\n",
    "    subset_input_dir = inputs_dir / f'subset_{subset_idx}'\n",
    "    command = (f'python {nerpa_dir / \"nerpa.py\"} '\n",
    "               f'--antismash-paths-file {subset_input_dir / \"antismash_paths.txt\"} '\n",
    "               f'--rban-json {nerpa_dir / \"data/mibig_norine_rban_preprocessed.json\"} '\n",
    "               f'--configs-dir {subset_input_dir / \"configs\"} '\n",
    "               '--process-hybrids '\n",
    "               '--max-num-matches-per-bgc 10 '\n",
    "               '--max-num-matches 1000 '\n",
    "               f'--output-dir {results_dir / f\"subset_{subset_idx}\"} '\n",
    "               f'--force-output-dir '\n",
    "               '--fast-matching '\n",
    "               '--threads 5 '\n",
    "               '--skip-molecule-drawing')\n",
    "    print(f'Running command: {command}')\n",
    "    subprocess.run(command, shell=True, check=True)"
   ],
   "id": "e642f9f376486fe8",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running command: python /home/ilianolhin/git/nerpa2/nerpa.py --antismash-paths-file /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_inputs/subset_0/antismash_paths.txt --rban-json /home/ilianolhin/git/nerpa2/data/mibig_norine_rban_preprocessed.json --configs-dir /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_inputs/subset_0/configs --process-hybrids --max-num-matches-per-bgc 10 --max-num-matches 1000 --output-dir /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_0 --force-output-dir --fast-matching --threads 5 --skip-molecule-drawing\n",
      "\n",
      "Started with command: /home/ilianolhin/git/nerpa2/nerpa.py --antismash-paths-file /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_inputs/subset_0/antismash_paths.txt --rban-json /home/ilianolhin/git/nerpa2/data/mibig_norine_rban_preprocessed.json --configs-dir /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_inputs/subset_0/configs --process-hybrids --max-num-matches-per-bgc 10 --max-num-matches 1000 --output-dir /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_0 --force-output-dir --fast-matching --threads 5 --skip-molecule-drawing\n",
      "\n",
      "System information:\n",
      "  OS: Linux-5.15.0-48-generic-x86_64-with-glibc2.35\n",
      "  Python version: 3.11\n",
      "  CPUs number: 12\n",
      "\n",
      "Started: 2025-06-11 17:55:45\n",
      "\n",
      "Logging to /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_0/nerpa.log\n",
      "\n",
      "======= Predicting BGC variants from antiSMASH results\n",
      "\n",
      "======= Done with Predicting BGC variants\n",
      "Getting BGC variants took 3.1515 seconds\n",
      "\n",
      "======= Constructing HMMs\n",
      "Constructing HMMs took 0.2607 seconds\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[17:55:50] Can't kekulize mol.  Unkekulized atoms: 6 7 8 11 12\n",
      "[17:55:50] non-ring atom 1 marked aromatic\n",
      "[17:55:50] non-ring atom 0 marked aromatic\n",
      "[17:55:50] Explicit valence for atom # 7 N, 4, is greater than permitted\n",
      "[17:55:50] Can't kekulize mol.  Unkekulized atoms: 2 3 4\n",
      "[17:55:50] Explicit valence for atom # 0 C, 5, is greater than permitted\n",
      "[17:55:50] non-ring atom 6 marked aromatic\n",
      "[17:55:50] non-ring atom 6 marked aromatic\n",
      "[17:55:50] Can't kekulize mol.  Unkekulized atoms: 6 7 8 10 11\n",
      "[17:55:50] non-ring atom 10 marked aromatic\n",
      "[17:55:50] non-ring atom 10 marked aromatic\n",
      "[17:55:50] non-ring atom 10 marked aromatic\n",
      "[17:55:50] non-ring atom 10 marked aromatic\n",
      "[17:55:50] non-ring atom 10 marked aromatic\n",
      "[17:55:50] non-ring atom 10 marked aromatic\n",
      "[17:55:50] non-ring atom 10 marked aromatic\n",
      "[17:55:50] non-ring atom 4 marked aromatic\n",
      "[17:55:50] non-ring atom 4 marked aromatic\n",
      "[17:55:50] non-ring atom 0 marked aromatic\n",
      "[17:55:50] Explicit valence for atom # 0 C, 5, is greater than permitted\n",
      "[17:55:50] Explicit valence for atom # 0 C, 5, is greater than permitted\n",
      "[17:55:50] non-ring atom 0 marked aromatic\n",
      "[17:55:50] non-ring atom 6 marked aromatic\n",
      "[17:55:50] non-ring atom 0 marked aromatic\n",
      "[17:55:50] Can't kekulize mol.  Unkekulized atoms: 2 3 4\n",
      "[17:55:50] Can't kekulize mol.  Unkekulized atoms: 0 5 6\n",
      "[17:55:50] Can't kekulize mol.  Unkekulized atoms: 0 5 6\n",
      "[17:55:50] Can't kekulize mol.  Unkekulized atoms: 3 4 5\n",
      "[17:55:50] Can't kekulize mol.  Unkekulized atoms: 0 5 6\n",
      "[17:55:50] non-ring atom 10 marked aromatic\n",
      "[17:55:50] Explicit valence for atom # 0 C, 5, is greater than permitted\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "=== Resolving NRP-PK hybrid monomers candidates\n",
      "\n",
      "== Running: java -jar /home/ilianolhin/git/nerpa2/external_tools/rBAN/rBAN-1.0.jar -inputFile /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_0/hybrid_monomers_input.json -outputFolder /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_0/ -outputFileName hybrid_monomers_output.json -monomersDB /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_0/rban_monomers_db.json\n",
      "\n",
      "  WARNING WRONG SMILES: 539_16 could not parse 'Nc1cc(Cl)ccc1(=O)O', a valid kekulé structure could not be assigned\n",
      "  WARNING WRONG SMILES: 540_6 could not parse 'Nc1cc(Cl)ccc1(=O)O', a valid kekulé structure could not be assigned\n",
      "  WARNING WRONG SMILES: 819_3 could not parse 'Nc1cc(Cl)ccc1(=O)O', a valid kekulé structure could not be assigned\n",
      "  Process completed successfully. Compounds analysed: 294\n",
      "\n",
      "== Done\n",
      "\n",
      "\n",
      "======= Done with Structures preprocessing with rBAN\n",
      "\n",
      "======= Processing rBAN output\n",
      "results will be in /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_0/NRP_Variants\n",
      "WARNING: Structure \"NOR01254\": unable to determine backbone sequence. Skipping \"NOR01254\".\n",
      "WARNING: Structure \"BGC0002531.0\": unable to determine backbone sequence. Skipping \"BGC0002531.0\".\n",
      "WARNING: Structure \"BGC0001635.0\": unable to determine backbone sequence. Skipping \"BGC0001635.0\".\n",
      "WARNING: Structure \"BGC0001502.0\": unable to determine backbone sequence. Skipping \"BGC0001502.0\".\n",
      "WARNING: Structure \"BGC0000326.0\": unable to determine backbone sequence. Skipping \"BGC0000326.0\".\n",
      "WARNING: Structure \"BGC0002892.0\": unable to determine backbone sequence. Skipping \"BGC0002892.0\".\n",
      "WARNING: Structure \"NOR00475\": unable to determine backbone sequence. Skipping \"NOR00475\".\n",
      "WARNING: Structure \"NOR00476\": unable to determine backbone sequence. Skipping \"NOR00476\".\n",
      "WARNING: Structure \"NOR00477\": unable to determine backbone sequence. Skipping \"NOR00477\".\n",
      "WARNING: Structure \"NOR00478\": unable to determine backbone sequence. Skipping \"NOR00478\".\n",
      "WARNING: Structure \"NOR00479\": unable to determine backbone sequence. Skipping \"NOR00479\".\n",
      "WARNING: Structure \"BGC0002073.0\": unable to determine backbone sequence. Skipping \"BGC0002073.0\".\n",
      "WARNING: Structure \"BGC0002314.0\": unable to determine backbone sequence. Skipping \"BGC0002314.0\".\n",
      "WARNING: Structure \"BGC0002430.1\": unable to determine backbone sequence. Skipping \"BGC0002430.1\".\n",
      "WARNING: Structure \"BGC0002367.0\": unable to determine backbone sequence. Skipping \"BGC0002367.0\".\n",
      "WARNING: Structure \"NOR00598\": unable to determine backbone sequence. Skipping \"NOR00598\".\n",
      "WARNING: Structure \"BGC0001250.0\": unable to determine backbone sequence. Skipping \"BGC0001250.0\".\n",
      "WARNING! Multiple hamiltonian paths found. Proceeding with the first one\n",
      "WARNING! Multiple hamiltonian paths found. Proceeding with the first one\n",
      "WARNING: Structure \"BGC0000371.0\": unable to determine backbone sequence. Skipping \"BGC0000371.0\".\n",
      "WARNING: Structure \"BGC0000371.1\": unable to determine backbone sequence. Skipping \"BGC0000371.1\".\n",
      "WARNING: Structure \"BGC0002378.1\": unable to determine backbone sequence. Skipping \"BGC0002378.1\".\n",
      "WARNING: Structure \"BGC0002378.0\": unable to determine backbone sequence. Skipping \"BGC0002378.0\".\n",
      "\n",
      "======= Done with Processing rBAN output\n",
      "Getting NRP variants took 7.0718 seconds\n",
      "\n",
      "======= Generating NRP linearizations\n",
      "Generating linearizations took 0.0083 seconds\n",
      "\n",
      "======= Nerpa matching\n",
      "Parsing HMMs from /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_0/hmms.json\n",
      "Parsing NRP linearizations from /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_0/hmms.json\n",
      "Matching 53 HMMs against 1549 NRP linearizations in 5 threads...\n",
      "Matching took 1.99 s\n",
      "Writing matches to /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_0/cpp_matching_output.json\n",
      "Done! Wrote 340 matches to /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_0/cpp_matching_output.json\n",
      "Running C++ matcher with command: /home/ilianolhin/git/nerpa2/build/hmm_nrp_matcher --hmms_json /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_0/hmms.json --nrps_json /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_0/nrp_linearizations.json --max_num_matches_per_bgc 10 --max_num_matches_per_nrp 0 --max_num_matches 1000 --threads 5 --output /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_0/cpp_matching_output.json\n",
      "Matching took 2.6383 seconds\n",
      "\n",
      "======= Reconstructing alignments for matches\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ilianolhin/git/nerpa2/src/matching/p_values_estimation.py:156: RuntimeWarning: overflow encountered in scalar add\n",
      "  dp[next_state][disc_new_path_lp] += number_of_paths\n",
      "/home/ilianolhin/git/nerpa2/src/matching/p_values_estimation.py:163: RuntimeWarning: overflow encountered in scalar add\n",
      "  dp[next_state][disc_new_path_lp] += number_of_paths\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "======= Writing results\n",
      "RESULTS:\n",
      "  Main report is saved to /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_0/report.tsv\n",
      "  HTML report is saved to /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_0/report.html\n",
      "  Detailed reports are saved to /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_0/matches_details\n",
      "Writing results took 4.5099 seconds\n",
      "  Log is saved to /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_0/nerpa.log\n",
      "\n",
      "Finished: 2025-06-11 17:56:11\n",
      "Elapsed time: 0:00:25.920359\n",
      "WARNINGs: 21\n",
      "\n",
      "Thank you for using Nerpa!\n",
      "Running command: python /home/ilianolhin/git/nerpa2/nerpa.py --antismash-paths-file /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_inputs/subset_1/antismash_paths.txt --rban-json /home/ilianolhin/git/nerpa2/data/mibig_norine_rban_preprocessed.json --configs-dir /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_inputs/subset_1/configs --process-hybrids --max-num-matches-per-bgc 10 --max-num-matches 1000 --output-dir /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_1 --force-output-dir --fast-matching --threads 5 --skip-molecule-drawing\n",
      "\n",
      "Started with command: /home/ilianolhin/git/nerpa2/nerpa.py --antismash-paths-file /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_inputs/subset_1/antismash_paths.txt --rban-json /home/ilianolhin/git/nerpa2/data/mibig_norine_rban_preprocessed.json --configs-dir /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_inputs/subset_1/configs --process-hybrids --max-num-matches-per-bgc 10 --max-num-matches 1000 --output-dir /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_1 --force-output-dir --fast-matching --threads 5 --skip-molecule-drawing\n",
      "\n",
      "System information:\n",
      "  OS: Linux-5.15.0-48-generic-x86_64-with-glibc2.35\n",
      "  Python version: 3.11\n",
      "  CPUs number: 12\n",
      "\n",
      "Started: 2025-06-11 17:56:13\n",
      "\n",
      "Logging to /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_1/nerpa.log\n",
      "\n",
      "======= Predicting BGC variants from antiSMASH results\n",
      "\n",
      "======= Done with Predicting BGC variants\n",
      "Getting BGC variants took 3.0184 seconds\n",
      "\n",
      "======= Constructing HMMs\n",
      "Constructing HMMs took 0.6557 seconds\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[17:56:18] Can't kekulize mol.  Unkekulized atoms: 6 7 8 11 12\n",
      "[17:56:18] non-ring atom 1 marked aromatic\n",
      "[17:56:18] non-ring atom 0 marked aromatic\n",
      "[17:56:19] Explicit valence for atom # 7 N, 4, is greater than permitted\n",
      "[17:56:19] Can't kekulize mol.  Unkekulized atoms: 2 3 4\n",
      "[17:56:19] Explicit valence for atom # 0 C, 5, is greater than permitted\n",
      "[17:56:19] non-ring atom 6 marked aromatic\n",
      "[17:56:19] non-ring atom 6 marked aromatic\n",
      "[17:56:19] Can't kekulize mol.  Unkekulized atoms: 6 7 8 10 11\n",
      "[17:56:19] non-ring atom 10 marked aromatic\n",
      "[17:56:19] non-ring atom 10 marked aromatic\n",
      "[17:56:19] non-ring atom 10 marked aromatic\n",
      "[17:56:19] non-ring atom 10 marked aromatic\n",
      "[17:56:19] non-ring atom 10 marked aromatic\n",
      "[17:56:19] non-ring atom 10 marked aromatic\n",
      "[17:56:19] non-ring atom 10 marked aromatic\n",
      "[17:56:19] non-ring atom 4 marked aromatic\n",
      "[17:56:19] non-ring atom 4 marked aromatic\n",
      "[17:56:19] non-ring atom 0 marked aromatic\n",
      "[17:56:19] Explicit valence for atom # 0 C, 5, is greater than permitted\n",
      "[17:56:19] Explicit valence for atom # 0 C, 5, is greater than permitted\n",
      "[17:56:19] non-ring atom 0 marked aromatic\n",
      "[17:56:19] non-ring atom 6 marked aromatic\n",
      "[17:56:19] non-ring atom 0 marked aromatic\n",
      "[17:56:19] Can't kekulize mol.  Unkekulized atoms: 2 3 4\n",
      "[17:56:19] Can't kekulize mol.  Unkekulized atoms: 0 5 6\n",
      "[17:56:19] Can't kekulize mol.  Unkekulized atoms: 0 5 6\n",
      "[17:56:19] Can't kekulize mol.  Unkekulized atoms: 3 4 5\n",
      "[17:56:19] Can't kekulize mol.  Unkekulized atoms: 0 5 6\n",
      "[17:56:19] non-ring atom 10 marked aromatic\n",
      "[17:56:19] Explicit valence for atom # 0 C, 5, is greater than permitted\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "=== Resolving NRP-PK hybrid monomers candidates\n",
      "\n",
      "== Running: java -jar /home/ilianolhin/git/nerpa2/external_tools/rBAN/rBAN-1.0.jar -inputFile /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_1/hybrid_monomers_input.json -outputFolder /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_1/ -outputFileName hybrid_monomers_output.json -monomersDB /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_1/rban_monomers_db.json\n",
      "\n",
      "  WARNING WRONG SMILES: 539_16 could not parse 'Nc1cc(Cl)ccc1(=O)O', a valid kekulé structure could not be assigned\n",
      "  WARNING WRONG SMILES: 540_6 could not parse 'Nc1cc(Cl)ccc1(=O)O', a valid kekulé structure could not be assigned\n",
      "  WARNING WRONG SMILES: 819_3 could not parse 'Nc1cc(Cl)ccc1(=O)O', a valid kekulé structure could not be assigned\n",
      "  Process completed successfully. Compounds analysed: 294\n",
      "\n",
      "== Done\n",
      "\n",
      "\n",
      "======= Done with Structures preprocessing with rBAN\n",
      "\n",
      "======= Processing rBAN output\n",
      "results will be in /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_1/NRP_Variants\n",
      "WARNING: Structure \"NOR01254\": unable to determine backbone sequence. Skipping \"NOR01254\".\n",
      "WARNING: Structure \"BGC0002531.0\": unable to determine backbone sequence. Skipping \"BGC0002531.0\".\n",
      "WARNING: Structure \"BGC0001635.0\": unable to determine backbone sequence. Skipping \"BGC0001635.0\".\n",
      "WARNING: Structure \"BGC0001502.0\": unable to determine backbone sequence. Skipping \"BGC0001502.0\".\n",
      "WARNING: Structure \"BGC0000326.0\": unable to determine backbone sequence. Skipping \"BGC0000326.0\".\n",
      "WARNING: Structure \"BGC0002892.0\": unable to determine backbone sequence. Skipping \"BGC0002892.0\".\n",
      "WARNING: Structure \"NOR00475\": unable to determine backbone sequence. Skipping \"NOR00475\".\n",
      "WARNING: Structure \"NOR00476\": unable to determine backbone sequence. Skipping \"NOR00476\".\n",
      "WARNING: Structure \"NOR00477\": unable to determine backbone sequence. Skipping \"NOR00477\".\n",
      "WARNING: Structure \"NOR00478\": unable to determine backbone sequence. Skipping \"NOR00478\".\n",
      "WARNING: Structure \"NOR00479\": unable to determine backbone sequence. Skipping \"NOR00479\".\n",
      "WARNING: Structure \"BGC0002073.0\": unable to determine backbone sequence. Skipping \"BGC0002073.0\".\n",
      "WARNING: Structure \"BGC0002314.0\": unable to determine backbone sequence. Skipping \"BGC0002314.0\".\n",
      "WARNING: Structure \"BGC0002430.1\": unable to determine backbone sequence. Skipping \"BGC0002430.1\".\n",
      "WARNING: Structure \"BGC0002367.0\": unable to determine backbone sequence. Skipping \"BGC0002367.0\".\n",
      "WARNING: Structure \"NOR00598\": unable to determine backbone sequence. Skipping \"NOR00598\".\n",
      "WARNING: Structure \"BGC0001250.0\": unable to determine backbone sequence. Skipping \"BGC0001250.0\".\n",
      "WARNING! Multiple hamiltonian paths found. Proceeding with the first one\n",
      "WARNING! Multiple hamiltonian paths found. Proceeding with the first one\n",
      "WARNING: Structure \"BGC0000371.0\": unable to determine backbone sequence. Skipping \"BGC0000371.0\".\n",
      "WARNING: Structure \"BGC0000371.1\": unable to determine backbone sequence. Skipping \"BGC0000371.1\".\n",
      "WARNING: Structure \"BGC0002378.1\": unable to determine backbone sequence. Skipping \"BGC0002378.1\".\n",
      "WARNING: Structure \"BGC0002378.0\": unable to determine backbone sequence. Skipping \"BGC0002378.0\".\n",
      "\n",
      "======= Done with Processing rBAN output\n",
      "Getting NRP variants took 10.4620 seconds\n",
      "\n",
      "======= Generating NRP linearizations\n",
      "Generating linearizations took 0.0088 seconds\n",
      "\n",
      "======= Nerpa matching\n",
      "Parsing HMMs from /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_1/hmms.json\n",
      "Parsing NRP linearizations from /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_1/hmms.json\n",
      "Matching 53 HMMs against 1549 NRP linearizations in 5 threads...\n",
      "Matching took 2.475 s\n",
      "Writing matches to /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_1/cpp_matching_output.json\n",
      "Done! Wrote 220 matches to /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_1/cpp_matching_output.json\n",
      "Running C++ matcher with command: /home/ilianolhin/git/nerpa2/build/hmm_nrp_matcher --hmms_json /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_1/hmms.json --nrps_json /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_1/nrp_linearizations.json --max_num_matches_per_bgc 10 --max_num_matches_per_nrp 0 --max_num_matches 1000 --threads 5 --output /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_1/cpp_matching_output.json\n",
      "Matching took 3.2813 seconds\n",
      "\n",
      "======= Reconstructing alignments for matches\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ilianolhin/git/nerpa2/src/matching/p_values_estimation.py:156: RuntimeWarning: overflow encountered in scalar add\n",
      "  dp[next_state][disc_new_path_lp] += number_of_paths\n",
      "/home/ilianolhin/git/nerpa2/src/matching/p_values_estimation.py:163: RuntimeWarning: overflow encountered in scalar add\n",
      "  dp[next_state][disc_new_path_lp] += number_of_paths\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "======= Writing results\n",
      "RESULTS:\n",
      "  Main report is saved to /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_1/report.tsv\n",
      "  HTML report is saved to /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_1/report.html\n",
      "  Detailed reports are saved to /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_1/matches_details\n",
      "Writing results took 3.9544 seconds\n",
      "  Log is saved to /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_1/nerpa.log\n",
      "\n",
      "Finished: 2025-06-11 17:56:41\n",
      "Elapsed time: 0:00:28.928256\n",
      "WARNINGs: 21\n",
      "\n",
      "Thank you for using Nerpa!\n",
      "Running command: python /home/ilianolhin/git/nerpa2/nerpa.py --antismash-paths-file /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_inputs/subset_2/antismash_paths.txt --rban-json /home/ilianolhin/git/nerpa2/data/mibig_norine_rban_preprocessed.json --configs-dir /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_inputs/subset_2/configs --process-hybrids --max-num-matches-per-bgc 10 --max-num-matches 1000 --output-dir /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_2 --force-output-dir --fast-matching --threads 5 --skip-molecule-drawing\n",
      "\n",
      "Started with command: /home/ilianolhin/git/nerpa2/nerpa.py --antismash-paths-file /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_inputs/subset_2/antismash_paths.txt --rban-json /home/ilianolhin/git/nerpa2/data/mibig_norine_rban_preprocessed.json --configs-dir /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_inputs/subset_2/configs --process-hybrids --max-num-matches-per-bgc 10 --max-num-matches 1000 --output-dir /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_2 --force-output-dir --fast-matching --threads 5 --skip-molecule-drawing\n",
      "\n",
      "System information:\n",
      "  OS: Linux-5.15.0-48-generic-x86_64-with-glibc2.35\n",
      "  Python version: 3.11\n",
      "  CPUs number: 12\n",
      "\n",
      "Started: 2025-06-11 17:56:43\n",
      "\n",
      "Logging to /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_2/nerpa.log\n",
      "\n",
      "======= Predicting BGC variants from antiSMASH results\n",
      "\n",
      "======= Done with Predicting BGC variants\n",
      "Getting BGC variants took 4.9726 seconds\n",
      "\n",
      "======= Constructing HMMs\n",
      "Constructing HMMs took 0.5003 seconds\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[17:56:51] Can't kekulize mol.  Unkekulized atoms: 6 7 8 11 12\n",
      "[17:56:51] non-ring atom 1 marked aromatic\n",
      "[17:56:51] non-ring atom 0 marked aromatic\n",
      "[17:56:51] Explicit valence for atom # 7 N, 4, is greater than permitted\n",
      "[17:56:51] Can't kekulize mol.  Unkekulized atoms: 2 3 4\n",
      "[17:56:51] Explicit valence for atom # 0 C, 5, is greater than permitted\n",
      "[17:56:51] non-ring atom 6 marked aromatic\n",
      "[17:56:51] non-ring atom 6 marked aromatic\n",
      "[17:56:51] Can't kekulize mol.  Unkekulized atoms: 6 7 8 10 11\n",
      "[17:56:51] non-ring atom 10 marked aromatic\n",
      "[17:56:51] non-ring atom 10 marked aromatic\n",
      "[17:56:51] non-ring atom 10 marked aromatic\n",
      "[17:56:51] non-ring atom 10 marked aromatic\n",
      "[17:56:51] non-ring atom 10 marked aromatic\n",
      "[17:56:51] non-ring atom 10 marked aromatic\n",
      "[17:56:51] non-ring atom 10 marked aromatic\n",
      "[17:56:51] non-ring atom 4 marked aromatic\n",
      "[17:56:51] non-ring atom 4 marked aromatic\n",
      "[17:56:51] non-ring atom 0 marked aromatic\n",
      "[17:56:51] Explicit valence for atom # 0 C, 5, is greater than permitted\n",
      "[17:56:51] Explicit valence for atom # 0 C, 5, is greater than permitted\n",
      "[17:56:51] non-ring atom 0 marked aromatic\n",
      "[17:56:51] non-ring atom 6 marked aromatic\n",
      "[17:56:51] non-ring atom 0 marked aromatic\n",
      "[17:56:51] Can't kekulize mol.  Unkekulized atoms: 2 3 4\n",
      "[17:56:51] Can't kekulize mol.  Unkekulized atoms: 0 5 6\n",
      "[17:56:51] Can't kekulize mol.  Unkekulized atoms: 0 5 6\n",
      "[17:56:51] Can't kekulize mol.  Unkekulized atoms: 3 4 5\n",
      "[17:56:51] Can't kekulize mol.  Unkekulized atoms: 0 5 6\n",
      "[17:56:51] non-ring atom 10 marked aromatic\n",
      "[17:56:51] Explicit valence for atom # 0 C, 5, is greater than permitted\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "=== Resolving NRP-PK hybrid monomers candidates\n",
      "\n",
      "== Running: java -jar /home/ilianolhin/git/nerpa2/external_tools/rBAN/rBAN-1.0.jar -inputFile /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_2/hybrid_monomers_input.json -outputFolder /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_2/ -outputFileName hybrid_monomers_output.json -monomersDB /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_2/rban_monomers_db.json\n",
      "\n",
      "  WARNING WRONG SMILES: 539_16 could not parse 'Nc1cc(Cl)ccc1(=O)O', a valid kekulé structure could not be assigned\n",
      "  WARNING WRONG SMILES: 540_6 could not parse 'Nc1cc(Cl)ccc1(=O)O', a valid kekulé structure could not be assigned\n",
      "  WARNING WRONG SMILES: 819_3 could not parse 'Nc1cc(Cl)ccc1(=O)O', a valid kekulé structure could not be assigned\n",
      "  Process completed successfully. Compounds analysed: 294\n",
      "\n",
      "== Done\n",
      "\n",
      "\n",
      "======= Done with Structures preprocessing with rBAN\n",
      "\n",
      "======= Processing rBAN output\n",
      "results will be in /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_2/NRP_Variants\n",
      "WARNING: Structure \"NOR01254\": unable to determine backbone sequence. Skipping \"NOR01254\".\n",
      "WARNING: Structure \"BGC0002531.0\": unable to determine backbone sequence. Skipping \"BGC0002531.0\".\n",
      "WARNING: Structure \"BGC0001635.0\": unable to determine backbone sequence. Skipping \"BGC0001635.0\".\n",
      "WARNING: Structure \"BGC0001502.0\": unable to determine backbone sequence. Skipping \"BGC0001502.0\".\n",
      "WARNING: Structure \"BGC0000326.0\": unable to determine backbone sequence. Skipping \"BGC0000326.0\".\n",
      "WARNING: Structure \"BGC0002892.0\": unable to determine backbone sequence. Skipping \"BGC0002892.0\".\n",
      "WARNING: Structure \"NOR00475\": unable to determine backbone sequence. Skipping \"NOR00475\".\n",
      "WARNING: Structure \"NOR00476\": unable to determine backbone sequence. Skipping \"NOR00476\".\n",
      "WARNING: Structure \"NOR00477\": unable to determine backbone sequence. Skipping \"NOR00477\".\n",
      "WARNING: Structure \"NOR00478\": unable to determine backbone sequence. Skipping \"NOR00478\".\n",
      "WARNING: Structure \"NOR00479\": unable to determine backbone sequence. Skipping \"NOR00479\".\n",
      "WARNING: Structure \"BGC0002073.0\": unable to determine backbone sequence. Skipping \"BGC0002073.0\".\n",
      "WARNING: Structure \"BGC0002314.0\": unable to determine backbone sequence. Skipping \"BGC0002314.0\".\n",
      "WARNING: Structure \"BGC0002430.1\": unable to determine backbone sequence. Skipping \"BGC0002430.1\".\n",
      "WARNING: Structure \"BGC0002367.0\": unable to determine backbone sequence. Skipping \"BGC0002367.0\".\n",
      "WARNING: Structure \"NOR00598\": unable to determine backbone sequence. Skipping \"NOR00598\".\n",
      "WARNING: Structure \"BGC0001250.0\": unable to determine backbone sequence. Skipping \"BGC0001250.0\".\n",
      "WARNING! Multiple hamiltonian paths found. Proceeding with the first one\n",
      "WARNING! Multiple hamiltonian paths found. Proceeding with the first one\n",
      "WARNING: Structure \"BGC0000371.0\": unable to determine backbone sequence. Skipping \"BGC0000371.0\".\n",
      "WARNING: Structure \"BGC0000371.1\": unable to determine backbone sequence. Skipping \"BGC0000371.1\".\n",
      "WARNING: Structure \"BGC0002378.1\": unable to determine backbone sequence. Skipping \"BGC0002378.1\".\n",
      "WARNING: Structure \"BGC0002378.0\": unable to determine backbone sequence. Skipping \"BGC0002378.0\".\n",
      "\n",
      "======= Done with Processing rBAN output\n",
      "Getting NRP variants took 9.9098 seconds\n",
      "\n",
      "======= Generating NRP linearizations\n",
      "Generating linearizations took 0.0095 seconds\n",
      "\n",
      "======= Nerpa matching\n",
      "Parsing HMMs from /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_2/hmms.json\n",
      "Parsing NRP linearizations from /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_2/hmms.json\n",
      "Matching 47 HMMs against 1549 NRP linearizations in 5 threads...\n",
      "Matching took 2.171 s\n",
      "Writing matches to /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_2/cpp_matching_output.json\n",
      "Done! Wrote 220 matches to /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_2/cpp_matching_output.json\n",
      "Running C++ matcher with command: /home/ilianolhin/git/nerpa2/build/hmm_nrp_matcher --hmms_json /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_2/hmms.json --nrps_json /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_2/nrp_linearizations.json --max_num_matches_per_bgc 10 --max_num_matches_per_nrp 0 --max_num_matches 1000 --threads 5 --output /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_2/cpp_matching_output.json\n",
      "Matching took 2.8647 seconds\n",
      "\n",
      "======= Reconstructing alignments for matches\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ilianolhin/git/nerpa2/src/matching/p_values_estimation.py:156: RuntimeWarning: overflow encountered in scalar add\n",
      "  dp[next_state][disc_new_path_lp] += number_of_paths\n",
      "/home/ilianolhin/git/nerpa2/src/matching/p_values_estimation.py:163: RuntimeWarning: overflow encountered in scalar add\n",
      "  dp[next_state][disc_new_path_lp] += number_of_paths\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "======= Writing results\n",
      "RESULTS:\n",
      "  Main report is saved to /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_2/report.tsv\n",
      "  HTML report is saved to /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_2/report.html\n",
      "  Detailed reports are saved to /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_2/matches_details\n",
      "Writing results took 3.8709 seconds\n",
      "  Log is saved to /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_2/nerpa.log\n",
      "\n",
      "Finished: 2025-06-11 17:57:13\n",
      "Elapsed time: 0:00:29.556972\n",
      "WARNINGs: 21\n",
      "\n",
      "Thank you for using Nerpa!\n",
      "Running command: python /home/ilianolhin/git/nerpa2/nerpa.py --antismash-paths-file /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_inputs/subset_3/antismash_paths.txt --rban-json /home/ilianolhin/git/nerpa2/data/mibig_norine_rban_preprocessed.json --configs-dir /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_inputs/subset_3/configs --process-hybrids --max-num-matches-per-bgc 10 --max-num-matches 1000 --output-dir /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_3 --force-output-dir --fast-matching --threads 5 --skip-molecule-drawing\n",
      "\n",
      "Started with command: /home/ilianolhin/git/nerpa2/nerpa.py --antismash-paths-file /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_inputs/subset_3/antismash_paths.txt --rban-json /home/ilianolhin/git/nerpa2/data/mibig_norine_rban_preprocessed.json --configs-dir /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_inputs/subset_3/configs --process-hybrids --max-num-matches-per-bgc 10 --max-num-matches 1000 --output-dir /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_3 --force-output-dir --fast-matching --threads 5 --skip-molecule-drawing\n",
      "\n",
      "System information:\n",
      "  OS: Linux-5.15.0-48-generic-x86_64-with-glibc2.35\n",
      "  Python version: 3.11\n",
      "  CPUs number: 12\n",
      "\n",
      "Started: 2025-06-11 17:57:14\n",
      "\n",
      "Logging to /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_3/nerpa.log\n",
      "\n",
      "======= Predicting BGC variants from antiSMASH results\n",
      "\n",
      "======= Done with Predicting BGC variants\n",
      "Getting BGC variants took 2.8652 seconds\n",
      "\n",
      "======= Constructing HMMs\n",
      "Constructing HMMs took 0.2924 seconds\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[17:57:19] Can't kekulize mol.  Unkekulized atoms: 6 7 8 11 12\n",
      "[17:57:19] non-ring atom 1 marked aromatic\n",
      "[17:57:19] non-ring atom 0 marked aromatic\n",
      "[17:57:19] Explicit valence for atom # 7 N, 4, is greater than permitted\n",
      "[17:57:19] Can't kekulize mol.  Unkekulized atoms: 2 3 4\n",
      "[17:57:19] Explicit valence for atom # 0 C, 5, is greater than permitted\n",
      "[17:57:19] non-ring atom 6 marked aromatic\n",
      "[17:57:19] non-ring atom 6 marked aromatic\n",
      "[17:57:19] Can't kekulize mol.  Unkekulized atoms: 6 7 8 10 11\n",
      "[17:57:19] non-ring atom 10 marked aromatic\n",
      "[17:57:19] non-ring atom 10 marked aromatic\n",
      "[17:57:19] non-ring atom 10 marked aromatic\n",
      "[17:57:19] non-ring atom 10 marked aromatic\n",
      "[17:57:19] non-ring atom 10 marked aromatic\n",
      "[17:57:19] non-ring atom 10 marked aromatic\n",
      "[17:57:19] non-ring atom 10 marked aromatic\n",
      "[17:57:19] non-ring atom 4 marked aromatic\n",
      "[17:57:19] non-ring atom 4 marked aromatic\n",
      "[17:57:19] non-ring atom 0 marked aromatic\n",
      "[17:57:19] Explicit valence for atom # 0 C, 5, is greater than permitted\n",
      "[17:57:19] Explicit valence for atom # 0 C, 5, is greater than permitted\n",
      "[17:57:19] non-ring atom 0 marked aromatic\n",
      "[17:57:19] non-ring atom 6 marked aromatic\n",
      "[17:57:19] non-ring atom 0 marked aromatic\n",
      "[17:57:19] Can't kekulize mol.  Unkekulized atoms: 2 3 4\n",
      "[17:57:19] Can't kekulize mol.  Unkekulized atoms: 0 5 6\n",
      "[17:57:19] Can't kekulize mol.  Unkekulized atoms: 0 5 6\n",
      "[17:57:19] Can't kekulize mol.  Unkekulized atoms: 3 4 5\n",
      "[17:57:19] Can't kekulize mol.  Unkekulized atoms: 0 5 6\n",
      "[17:57:19] non-ring atom 10 marked aromatic\n",
      "[17:57:19] Explicit valence for atom # 0 C, 5, is greater than permitted\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "=== Resolving NRP-PK hybrid monomers candidates\n",
      "\n",
      "== Running: java -jar /home/ilianolhin/git/nerpa2/external_tools/rBAN/rBAN-1.0.jar -inputFile /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_3/hybrid_monomers_input.json -outputFolder /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_3/ -outputFileName hybrid_monomers_output.json -monomersDB /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_3/rban_monomers_db.json\n",
      "\n",
      "  WARNING WRONG SMILES: 539_16 could not parse 'Nc1cc(Cl)ccc1(=O)O', a valid kekulé structure could not be assigned\n",
      "  WARNING WRONG SMILES: 540_6 could not parse 'Nc1cc(Cl)ccc1(=O)O', a valid kekulé structure could not be assigned\n",
      "  WARNING WRONG SMILES: 819_3 could not parse 'Nc1cc(Cl)ccc1(=O)O', a valid kekulé structure could not be assigned\n",
      "  Process completed successfully. Compounds analysed: 294\n",
      "\n",
      "== Done\n",
      "\n",
      "\n",
      "======= Done with Structures preprocessing with rBAN\n",
      "\n",
      "======= Processing rBAN output\n",
      "results will be in /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_3/NRP_Variants\n",
      "WARNING: Structure \"NOR01254\": unable to determine backbone sequence. Skipping \"NOR01254\".\n",
      "WARNING: Structure \"BGC0002531.0\": unable to determine backbone sequence. Skipping \"BGC0002531.0\".\n",
      "WARNING: Structure \"BGC0001635.0\": unable to determine backbone sequence. Skipping \"BGC0001635.0\".\n",
      "WARNING: Structure \"BGC0001502.0\": unable to determine backbone sequence. Skipping \"BGC0001502.0\".\n",
      "WARNING: Structure \"BGC0000326.0\": unable to determine backbone sequence. Skipping \"BGC0000326.0\".\n",
      "WARNING: Structure \"BGC0002892.0\": unable to determine backbone sequence. Skipping \"BGC0002892.0\".\n",
      "WARNING: Structure \"NOR00475\": unable to determine backbone sequence. Skipping \"NOR00475\".\n",
      "WARNING: Structure \"NOR00476\": unable to determine backbone sequence. Skipping \"NOR00476\".\n",
      "WARNING: Structure \"NOR00477\": unable to determine backbone sequence. Skipping \"NOR00477\".\n",
      "WARNING: Structure \"NOR00478\": unable to determine backbone sequence. Skipping \"NOR00478\".\n",
      "WARNING: Structure \"NOR00479\": unable to determine backbone sequence. Skipping \"NOR00479\".\n",
      "WARNING: Structure \"BGC0002073.0\": unable to determine backbone sequence. Skipping \"BGC0002073.0\".\n",
      "WARNING: Structure \"BGC0002314.0\": unable to determine backbone sequence. Skipping \"BGC0002314.0\".\n",
      "WARNING: Structure \"BGC0002430.1\": unable to determine backbone sequence. Skipping \"BGC0002430.1\".\n",
      "WARNING: Structure \"BGC0002367.0\": unable to determine backbone sequence. Skipping \"BGC0002367.0\".\n",
      "WARNING: Structure \"NOR00598\": unable to determine backbone sequence. Skipping \"NOR00598\".\n",
      "WARNING: Structure \"BGC0001250.0\": unable to determine backbone sequence. Skipping \"BGC0001250.0\".\n",
      "WARNING! Multiple hamiltonian paths found. Proceeding with the first one\n",
      "WARNING! Multiple hamiltonian paths found. Proceeding with the first one\n",
      "WARNING: Structure \"BGC0000371.0\": unable to determine backbone sequence. Skipping \"BGC0000371.0\".\n",
      "WARNING: Structure \"BGC0000371.1\": unable to determine backbone sequence. Skipping \"BGC0000371.1\".\n",
      "WARNING: Structure \"BGC0002378.1\": unable to determine backbone sequence. Skipping \"BGC0002378.1\".\n",
      "WARNING: Structure \"BGC0002378.0\": unable to determine backbone sequence. Skipping \"BGC0002378.0\".\n",
      "\n",
      "======= Done with Processing rBAN output\n",
      "Getting NRP variants took 7.0803 seconds\n",
      "\n",
      "======= Generating NRP linearizations\n",
      "Generating linearizations took 0.0094 seconds\n",
      "\n",
      "======= Nerpa matching\n",
      "Parsing HMMs from /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_3/hmms.json\n",
      "Parsing NRP linearizations from /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_3/hmms.json\n",
      "Matching 52 HMMs against 1549 NRP linearizations in 5 threads...\n",
      "Matching took 1.846 s\n",
      "Writing matches to /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_3/cpp_matching_output.json\n",
      "Done! Wrote 270 matches to /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_3/cpp_matching_output.json\n",
      "Running C++ matcher with command: /home/ilianolhin/git/nerpa2/build/hmm_nrp_matcher --hmms_json /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_3/hmms.json --nrps_json /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_3/nrp_linearizations.json --max_num_matches_per_bgc 10 --max_num_matches_per_nrp 0 --max_num_matches 1000 --threads 5 --output /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_3/cpp_matching_output.json\n",
      "Matching took 2.6061 seconds\n",
      "\n",
      "======= Reconstructing alignments for matches\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ilianolhin/git/nerpa2/src/matching/p_values_estimation.py:156: RuntimeWarning: overflow encountered in scalar add\n",
      "  dp[next_state][disc_new_path_lp] += number_of_paths\n",
      "/home/ilianolhin/git/nerpa2/src/matching/p_values_estimation.py:163: RuntimeWarning: overflow encountered in scalar add\n",
      "  dp[next_state][disc_new_path_lp] += number_of_paths\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "======= Writing results\n",
      "RESULTS:\n",
      "  Main report is saved to /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_3/report.tsv\n",
      "  HTML report is saved to /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_3/report.html\n",
      "  Detailed reports are saved to /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_3/matches_details\n",
      "Writing results took 4.4987 seconds\n",
      "  Log is saved to /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_3/nerpa.log\n",
      "\n",
      "Finished: 2025-06-11 17:57:39\n",
      "Elapsed time: 0:00:24.552692\n",
      "WARNINGs: 21\n",
      "\n",
      "Thank you for using Nerpa!\n",
      "Running command: python /home/ilianolhin/git/nerpa2/nerpa.py --antismash-paths-file /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_inputs/subset_4/antismash_paths.txt --rban-json /home/ilianolhin/git/nerpa2/data/mibig_norine_rban_preprocessed.json --configs-dir /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_inputs/subset_4/configs --process-hybrids --max-num-matches-per-bgc 10 --max-num-matches 1000 --output-dir /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_4 --force-output-dir --fast-matching --threads 5 --skip-molecule-drawing\n",
      "\n",
      "Started with command: /home/ilianolhin/git/nerpa2/nerpa.py --antismash-paths-file /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_inputs/subset_4/antismash_paths.txt --rban-json /home/ilianolhin/git/nerpa2/data/mibig_norine_rban_preprocessed.json --configs-dir /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_inputs/subset_4/configs --process-hybrids --max-num-matches-per-bgc 10 --max-num-matches 1000 --output-dir /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_4 --force-output-dir --fast-matching --threads 5 --skip-molecule-drawing\n",
      "\n",
      "System information:\n",
      "  OS: Linux-5.15.0-48-generic-x86_64-with-glibc2.35\n",
      "  Python version: 3.11\n",
      "  CPUs number: 12\n",
      "\n",
      "Started: 2025-06-11 17:57:41\n",
      "\n",
      "Logging to /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_4/nerpa.log\n",
      "\n",
      "======= Predicting BGC variants from antiSMASH results\n",
      "\n",
      "======= Done with Predicting BGC variants\n",
      "Getting BGC variants took 4.0168 seconds\n",
      "\n",
      "======= Constructing HMMs\n",
      "Constructing HMMs took 0.4286 seconds\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[17:57:46] Can't kekulize mol.  Unkekulized atoms: 6 7 8 11 12\n",
      "[17:57:46] non-ring atom 1 marked aromatic\n",
      "[17:57:46] non-ring atom 0 marked aromatic\n",
      "[17:57:46] Explicit valence for atom # 7 N, 4, is greater than permitted\n",
      "[17:57:46] Can't kekulize mol.  Unkekulized atoms: 2 3 4\n",
      "[17:57:46] Explicit valence for atom # 0 C, 5, is greater than permitted\n",
      "[17:57:46] non-ring atom 6 marked aromatic\n",
      "[17:57:46] non-ring atom 6 marked aromatic\n",
      "[17:57:46] Can't kekulize mol.  Unkekulized atoms: 6 7 8 10 11\n",
      "[17:57:46] non-ring atom 10 marked aromatic\n",
      "[17:57:46] non-ring atom 10 marked aromatic\n",
      "[17:57:46] non-ring atom 10 marked aromatic\n",
      "[17:57:46] non-ring atom 10 marked aromatic\n",
      "[17:57:46] non-ring atom 10 marked aromatic\n",
      "[17:57:46] non-ring atom 10 marked aromatic\n",
      "[17:57:46] non-ring atom 10 marked aromatic\n",
      "[17:57:46] non-ring atom 4 marked aromatic\n",
      "[17:57:46] non-ring atom 4 marked aromatic\n",
      "[17:57:46] non-ring atom 0 marked aromatic\n",
      "[17:57:46] Explicit valence for atom # 0 C, 5, is greater than permitted\n",
      "[17:57:46] Explicit valence for atom # 0 C, 5, is greater than permitted\n",
      "[17:57:46] non-ring atom 0 marked aromatic\n",
      "[17:57:46] non-ring atom 6 marked aromatic\n",
      "[17:57:46] non-ring atom 0 marked aromatic\n",
      "[17:57:46] Can't kekulize mol.  Unkekulized atoms: 2 3 4\n",
      "[17:57:46] Can't kekulize mol.  Unkekulized atoms: 0 5 6\n",
      "[17:57:46] Can't kekulize mol.  Unkekulized atoms: 0 5 6\n",
      "[17:57:46] Can't kekulize mol.  Unkekulized atoms: 3 4 5\n",
      "[17:57:46] Can't kekulize mol.  Unkekulized atoms: 0 5 6\n",
      "[17:57:47] non-ring atom 10 marked aromatic\n",
      "[17:57:47] Explicit valence for atom # 0 C, 5, is greater than permitted\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "=== Resolving NRP-PK hybrid monomers candidates\n",
      "\n",
      "== Running: java -jar /home/ilianolhin/git/nerpa2/external_tools/rBAN/rBAN-1.0.jar -inputFile /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_4/hybrid_monomers_input.json -outputFolder /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_4/ -outputFileName hybrid_monomers_output.json -monomersDB /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_4/rban_monomers_db.json\n",
      "\n",
      "  WARNING WRONG SMILES: 539_16 could not parse 'Nc1cc(Cl)ccc1(=O)O', a valid kekulé structure could not be assigned\n",
      "  WARNING WRONG SMILES: 540_6 could not parse 'Nc1cc(Cl)ccc1(=O)O', a valid kekulé structure could not be assigned\n",
      "  WARNING WRONG SMILES: 819_3 could not parse 'Nc1cc(Cl)ccc1(=O)O', a valid kekulé structure could not be assigned\n",
      "  Process completed successfully. Compounds analysed: 294\n",
      "\n",
      "== Done\n",
      "\n",
      "\n",
      "======= Done with Structures preprocessing with rBAN\n",
      "\n",
      "======= Processing rBAN output\n",
      "results will be in /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_4/NRP_Variants\n",
      "WARNING: Structure \"NOR01254\": unable to determine backbone sequence. Skipping \"NOR01254\".\n",
      "WARNING: Structure \"BGC0002531.0\": unable to determine backbone sequence. Skipping \"BGC0002531.0\".\n",
      "WARNING: Structure \"BGC0001635.0\": unable to determine backbone sequence. Skipping \"BGC0001635.0\".\n",
      "WARNING: Structure \"BGC0001502.0\": unable to determine backbone sequence. Skipping \"BGC0001502.0\".\n",
      "WARNING: Structure \"BGC0000326.0\": unable to determine backbone sequence. Skipping \"BGC0000326.0\".\n",
      "WARNING: Structure \"BGC0002892.0\": unable to determine backbone sequence. Skipping \"BGC0002892.0\".\n",
      "WARNING: Structure \"NOR00475\": unable to determine backbone sequence. Skipping \"NOR00475\".\n",
      "WARNING: Structure \"NOR00476\": unable to determine backbone sequence. Skipping \"NOR00476\".\n",
      "WARNING: Structure \"NOR00477\": unable to determine backbone sequence. Skipping \"NOR00477\".\n",
      "WARNING: Structure \"NOR00478\": unable to determine backbone sequence. Skipping \"NOR00478\".\n",
      "WARNING: Structure \"NOR00479\": unable to determine backbone sequence. Skipping \"NOR00479\".\n",
      "WARNING: Structure \"BGC0002073.0\": unable to determine backbone sequence. Skipping \"BGC0002073.0\".\n",
      "WARNING: Structure \"BGC0002314.0\": unable to determine backbone sequence. Skipping \"BGC0002314.0\".\n",
      "WARNING: Structure \"BGC0002430.1\": unable to determine backbone sequence. Skipping \"BGC0002430.1\".\n",
      "WARNING: Structure \"BGC0002367.0\": unable to determine backbone sequence. Skipping \"BGC0002367.0\".\n",
      "WARNING: Structure \"NOR00598\": unable to determine backbone sequence. Skipping \"NOR00598\".\n",
      "WARNING: Structure \"BGC0001250.0\": unable to determine backbone sequence. Skipping \"BGC0001250.0\".\n",
      "WARNING! Multiple hamiltonian paths found. Proceeding with the first one\n",
      "WARNING! Multiple hamiltonian paths found. Proceeding with the first one\n",
      "WARNING: Structure \"BGC0000371.0\": unable to determine backbone sequence. Skipping \"BGC0000371.0\".\n",
      "WARNING: Structure \"BGC0000371.1\": unable to determine backbone sequence. Skipping \"BGC0000371.1\".\n",
      "WARNING: Structure \"BGC0002378.1\": unable to determine backbone sequence. Skipping \"BGC0002378.1\".\n",
      "WARNING: Structure \"BGC0002378.0\": unable to determine backbone sequence. Skipping \"BGC0002378.0\".\n",
      "\n",
      "======= Done with Processing rBAN output\n",
      "Getting NRP variants took 6.7001 seconds\n",
      "\n",
      "======= Generating NRP linearizations\n",
      "Generating linearizations took 0.0091 seconds\n",
      "\n",
      "======= Nerpa matching\n",
      "Parsing HMMs from /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_4/hmms.json\n",
      "Parsing NRP linearizations from /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_4/hmms.json\n",
      "Matching 66 HMMs against 1549 NRP linearizations in 5 threads...\n",
      "Matching took 2.759 s\n",
      "Writing matches to /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_4/cpp_matching_output.json\n",
      "Done! Wrote 310 matches to /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_4/cpp_matching_output.json\n",
      "Running C++ matcher with command: /home/ilianolhin/git/nerpa2/build/hmm_nrp_matcher --hmms_json /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_4/hmms.json --nrps_json /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_4/nrp_linearizations.json --max_num_matches_per_bgc 10 --max_num_matches_per_nrp 0 --max_num_matches 1000 --threads 5 --output /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_4/cpp_matching_output.json\n",
      "Matching took 3.5702 seconds\n",
      "\n",
      "======= Reconstructing alignments for matches\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ilianolhin/git/nerpa2/src/matching/p_values_estimation.py:156: RuntimeWarning: overflow encountered in scalar add\n",
      "  dp[next_state][disc_new_path_lp] += number_of_paths\n",
      "/home/ilianolhin/git/nerpa2/src/matching/p_values_estimation.py:163: RuntimeWarning: overflow encountered in scalar add\n",
      "  dp[next_state][disc_new_path_lp] += number_of_paths\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "======= Writing results\n",
      "RESULTS:\n",
      "  Main report is saved to /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_4/report.tsv\n",
      "  HTML report is saved to /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_4/report.html\n",
      "  Detailed reports are saved to /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_4/matches_details\n",
      "Writing results took 4.4717 seconds\n",
      "  Log is saved to /home/ilianolhin/git/nerpa2/benchmarking/cross_validation_results/subset_4/nerpa.log\n",
      "\n",
      "Finished: 2025-06-11 17:58:09\n",
      "Elapsed time: 0:00:28.085012\n",
      "WARNINGs: 21\n",
      "\n",
      "Thank you for using Nerpa!\n"
     ]
    }
   ],
   "execution_count": 26
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
